dd-trace/src/telemetry.rs

```diff
@@ -0,0 +1,265 @@
+        self.send_stop()
+    }
+}
+
+struct Telemetry {
+    handle: Option<Box<dyn TelemetryHandle>>,
+    enabled: bool,
+    log_collection_enabled: bool,
+}
+
+pub fn init_telemetry(
+    config: &Config,
+    service_name: Option<String>,
+    custom_handle: Option<Box<dyn TelemetryHandle>>,
+) {
+    let _guard = INIT_TELEMETRY_LOCK.lock().unwrap();
+
+    if let Some(telemetry) = TELEMETRY.get() {
+        dd_debug!("Updating already initialized telemetry");
+
+        let mut telemetry = telemetry.lock().unwrap();
+        telemetry.enabled = config.telemetry_enabled();
+        telemetry.log_collection_enabled = config.telemetry_log_collection_enabled();
+
+        return;
+    } else if !config.telemetry_enabled() {
+        dd_info!("Telemetry not enabled");
+        return;
+    }
+
+    let handle: Option<Box<dyn TelemetryHandle>> = if custom_handle.is_none() {
+        let mut builder = worker::TelemetryWorkerBuilder::new(
+            "127.0.0.1".to_string(), // FIXME
+            service_name.unwrap_or(config.service().to_string()),
+            config.language().to_string(),
+            config.language_version().to_string(),
+            config.tracer_version().to_string(),
+        );
+        builder.config = ddtelemetry::config::Config::from_env();
+        builder.config.telemetry_heartbeat_interval =
+            Duration::from_secs_f64(config.telemetry_heartbeat_interval());
+        // builder.config.debug_enabled = true;
+
+        match builder.run() {
+            Ok(handle) => Some(Box::new(handle)),
+            Err(err) => {
+                dd_error!("Error initializing telemetry worker: {err:?}");
+                None
+            }
+        }
+    } else {
+        custom_handle
+    };
+
+    if let Some(ref handle) = handle {
+        handle.send_start().ok();
+    };
+
+    if TELEMETRY
+        .set(Arc::new(Mutex::new(Telemetry {
+            handle,
+            enabled: config.telemetry_enabled(),
+            log_collection_enabled: config.telemetry_log_collection_enabled(),
+        })))
+        .is_err()
+    {
+        dd_error!("Error initializing telemetry");
+    }
+}
+
+pub fn stop_telemetry() {
+    if let Some(telemetry) = TELEMETRY.get() {
+        if let Ok(telemetry) = telemetry.lock() {
+            if let Some(ref handle) = telemetry.handle {
+                dd_info!("Stopping telemetry");
+                handle.send_stop().ok();
+            }
+        }
+    }
+}
+
+// message should be a template and must avoid dynamic messages
+pub fn add_log_error<I: Into<String>>(message: I, stack: Option<String>) {
+    if let Some(telemetry) = TELEMETRY.get() {
+        if let Ok(telemetry) = telemetry.lock() {
+            if telemetry.enabled && telemetry.log_collection_enabled {
+                if let Some(handle) = &telemetry.handle {
+                    handle.add_error_log(message.into(), stack).ok();
+                }
+            }
+        }
+    }
```

Comment:
Nit but I prefer early return
```suggestion
    let Some(telemetry) = TELEMETRY.get() else { return; }
    let Ok(telemetry) = telemetry.lock() else { return; }
    if !telemetry.enabled || !telemetry.log_collection_enabled {
        return; 
    }
    let Some(handle) = &telemetry.handle else {
        return;
    }
    handle.add_error_log(message.into(), stack).ok();
```